test/testdrive/github-15748.td

```diff
@@ -0,0 +1,161 @@
+    }
+  },
+  {
+    "type": "record",
+    "name": "progress",
+    "namespace": "com.materialize.cdc",
+    "fields": [
+      {
+        "name": "lower",
+        "type": {
+          "type": "array",
+          "items": "long"
+        }
+      },
+      {
+        "name": "upper",
+        "type": {
+          "type": "array",
+          "items": "long"
+        }
+      },
+      {
+        "name": "counts",
+        "type": {
+          "type": "array",
+          "items": {
+            "type": "record",
+            "name": "counts",
+            "fields": [
+              {
+                "name": "time",
+                "type": "long"
+              },
+              {
+                "name": "count",
+                "type": "long"
+              }
+            ]
+          }
+        }
+      }
+    ]
+  }
+  ]
+
+
+#
+# Insert some rows and then delete, upsert and insert even more in the same timestamp
+#
+
+$ kafka-create-topic topic=topic1
+> CREATE SOURCE source1
+  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-topic1-${testdrive.seed}')
+  FORMAT AVRO USING SCHEMA '${schema}' ENVELOPE MATERIALIZE
+> CREATE SINK sink1 FROM source1
+  INTO KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-sink1-${testdrive.seed}') KEY (a) NOT ENFORCED
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn ENVELOPE UPSERT
+
+# We start with 10 records
+
+$ kafka-ingest format=avro topic=topic1 schema=${schema}
+{"array":[{"data":{"a":10,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":9,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":8,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":7,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":6,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":5,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":4,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":2,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":3,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":1,"b":1},"time":1,"diff":1}]}
+
+# All of the below happens in a single timestamp "time":2
+
+# Delete 2 records
+$ kafka-ingest format=avro topic=topic1 schema=${schema}
+{"array":[{"data":{"a":7,"b":1},"time":2,"diff":-1}]}
+{"array":[{"data":{"a":3,"b":1},"time":2,"diff":-1}]}
+
+# Upsert 2 records
+$ kafka-ingest format=avro topic=topic1 schema=${schema}
+{"array":[{"data":{"a":8,"b":1},"time":2,"diff":-1}]}
+{"array":[{"data":{"a":2,"b":1},"time":2,"diff":-1}]}
+{"array":[{"data":{"a":8,"b":8},"time":2,"diff":1}]}
+{"array":[{"data":{"a":2,"b":2},"time":2,"diff":1}]}
+
+# Insert 2 records
+$ kafka-ingest format=avro topic=topic1 schema=${schema}
+{"array":[{"data":{"a":0,"b":0},"time":2,"diff":1}]}
+{"array":[{"data":{"a":15,"b":15},"time":2,"diff":1}]}
+
+# Emit the progress
+$ kafka-ingest format=avro topic=topic1 schema=${schema}
+{"com.materialize.cdc.progress":{"lower":[0],"upper":[10],"counts":[{"time":1,"count":10}, {"time":2,"count":8}]}}
+
+# Because of sort-messages=true, the deletions must be listed first
```

Comment:
This test doesn't use `sort-messages=true`!

(In general I think we should probably be sorting here... these tests seem like they'd be sensitive to legal changes in the grouping logic or the kafka publishing.)

Reply:
Yes sorry the comment needs to be removed

Reply:
Comment removed! @bkirwi I couldn't figure out how to get `sort-messages` to do what I want—how do you verify that all of ts1 is emitted before all of ts2, for example? So going to merge as is.

Reply:
As a consumer you're not necessarily guaranteed to see all of `ts1` before `ts2`, depending on how fast you make progress on which partitions. I suppose it's arguable whether you'd rather overspecify or underspecify behaviour for stuff like this, though...

Possibly we should enhance kafka-verify to be able to more precisely state test properties. But we should probably make some headway on figuring out exactly what we want to promise users first...

Reply:
> As a consumer you're not necessarily guaranteed to see all of ts1 before ts2, depending on how fast you make progress on which partitions.

But if it's all one partition we guarantee in-order emission, right?

> Possibly we should enhance kafka-verify to be able to more precisely state test properties. But we should probably make some headway on figuring out exactly what we want to promise users first...

:1000: agree on both counts!